workflow/controller/operator.go

```diff
@@ -1445,10 +1445,10 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 		}
 	}
 
-	// We cannot fail the node until the wait container is finished (unless any init container has failed) because it may be busy saving outputs, and these
+	// We cannot fail the node until the wait container is finished (unless any init container has failed or wait contain is still Pending) because it may be busy saving outputs, and these
 	// would not get captured successfully.
 	for _, c := range pod.Status.ContainerStatuses {
-		if (c.Name == common.WaitContainerName && c.State.Terminated == nil && new.Phase.Completed()) && !initContainerFailed {
+		if (c.Name == common.WaitContainerName && c.State.Terminated == nil && c.State.Running != nil && new.Phase.Completed()) && !initContainerFailed {
```

Comment:
So, I was actually saying we can use `c.State.Running != nil` and thus remove `c.State.Terminated == nil` per this description of `ContainerState`:

```
// ContainerState holds a possible state of container.
// Only one of its members may be specified.
// If none of them is specified, the default one is ContainerStateWaiting.
type ContainerState struct {
	// Details about a waiting container
	// +optional
	Waiting *ContainerStateWaiting `json:"waiting,omitempty" protobuf:"bytes,1,opt,name=waiting"`
	// Details about a running container
	// +optional
	Running *ContainerStateRunning `json:"running,omitempty" protobuf:"bytes,2,opt,name=running"`
	// Details about a terminated container
	// +optional
	Terminated *ContainerStateTerminated `json:"terminated,omitempty" protobuf:"bytes,3,opt,name=terminated"`
}
```

The comment says "Only one of its members may be specified."

Reply:
Also, I'm thinking about whether we really need `!initContainerFailed`. If the wait container is running then the Init Container must not have failed, would it? (It may be worth trying to understand why that was put in there)

Reply:
Good catch, according to this description we may indeed only need "Running != nil".